src/OpenTK.Mathematics/Vector/Vector2i.cs

```diff
@@ -0,0 +1,544 @@
+        /// <param name="y">The Y component of the Vector2i.</param>
+        public Vector2i(int x, int y)
+        {
+            X = x;
+            Y = y;
+        }
+
+        /// <summary>
+        /// Gets or sets the value at the index of the vector.
+        /// </summary>
+        /// <param name="index">The index of the component from the vector.</param>
+        /// <exception cref="IndexOutOfRangeException">Thrown if the index is less than 0 or greater than 1.</exception>
+        public int this[int index]
+        {
+            get
+            {
+                if (index == 0)
+                {
+                    return X;
+                }
+
+                if (index == 1)
+                {
+                    return Y;
+                }
+
+                throw new IndexOutOfRangeException("You tried to access this vector at index: " + index);
+            }
+
+            set
+            {
+                if (index == 0)
+                {
+                    X = value;
+                }
+                else if (index == 1)
+                {
+                    Y = value;
+                }
+                else
+                {
+                    throw new IndexOutOfRangeException("You tried to set this vector at index: " + index);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the perpendicular vector on the right side of this vector.
+        /// </summary>
+        public Vector2i PerpendicularRight => new Vector2i(Y, -X);
+
+        /// <summary>
+        /// Gets the perpendicular vector on the left side of this vector.
+        /// </summary>
+        public Vector2i PerpendicularLeft => new Vector2i(-Y, X);
+
+        /// <summary>
+        /// Defines a unit-length <see cref="Vector2i"/> that points towards the X-axis.
+        /// </summary>
+        public static readonly Vector2i UnitX = new Vector2i(1, 0);
+
+        /// <summary>
+        /// Defines a unit-length <see cref="Vector2i"/> that points towards the Y-axis.
+        /// </summary>
+        public static readonly Vector2i UnitY = new Vector2i(0, 1);
+
+        /// <summary>
+        /// Defines a zero-length <see cref="Vector2i"/>.
+        /// </summary>
+        public static readonly Vector2i Zero = new Vector2i(0, 0);
+
+        /// <summary>
+        /// Defines an instance with all components set to 1.
+        /// </summary>
+        public static readonly Vector2i One = new Vector2i(1, 1);
+
+        /// <summary>
+        /// Defines the size of the <see cref="Vector2i"/> struct in bytes.
+        /// </summary>
+        public static readonly int SizeInBytes = Marshal.SizeOf<Vector2i>();
+
+        /// <summary>
+        /// Adds two vectors.
+        /// </summary>
+        /// <param name="a">Left operand.</param>
+        /// <param name="b">Right operand.</param>
+        /// <returns>Result of operation.</returns>
+        public static Vector2i Add(Vector2i a, Vector2i b)
```

Comment:
Can we annotate by-value methods with the [Pure](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.contracts.pureattribute?view=netframework-4.8) attribute?

This is something to look at for the rest of the vector/math library.